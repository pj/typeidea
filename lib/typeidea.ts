import * as crypto from 'crypto';
import {compile} from 'ejs';
import * as fs from 'fs';
import * as prettier from 'prettier';
import * as action from './action';

type Hash = string | null;

export function hashAction(
  action: action.Action,
  previous: string | null
) {
  const hash = crypto.createHash('sha256');

  if (previous !== null) {
    hash.update(previous);
  }

  const fieldsToHash = action.fieldsToHash();
  hash.update(fieldsToHash);

  return hash.digest('hex');
}

/**
* Take a list of types and verify the types and generate any empty types.
*/
export function hashActions(
  hashables: Array<action.Action>,
  requireHashes: boolean
): Array<[number, string]> {
  const newHashes: Array<[number, string]> = [];
  let previousHash = null;
  let createHashes = false;
  for (let n = 0; n < hashables.length; n++) {
    let hashable = hashables[n];
    if (createHashes) {
      if (hashable.hash !== null && hashable.hash !== undefined) {
        throw new Error(`Hash entered when there were previous items without hashes index: ${n} action: ${hashable}`);
      }
      previousHash = hashAction(hashable, previousHash);
      newHashes.push([n, previousHash]);
    } else {
      if (hashable.hash === null || hashable.hash === undefined) {
        if (requireHashes) {
          throw new Error(`Unhashed action found at item ${n} action: ${hashable}`);
        }
        previousHash = hashAction(hashable, previousHash);
        newHashes.push([n, previousHash]);
        createHashes = true;
      } else {
        const expectedHash = hashAction(hashable, previousHash);
        if (expectedHash !== hashable.hash) {
          throw new Error(`Invalid hash at item ${n} expected ${expectedHash} got ${hashable.hash} object: ${hashable}`)
        }
        previousHash = hashable.hash;
      }
    }
  }

  return newHashes;
}

export function addHashes(
  unhashedType: action.Action[],
  hashes: Array<[number, string]>,
  hashTo: number | null
) {
  if (hashes.length === 0) {
    return unhashedType;
  }
  const hashed = unhashedType.slice();
  if (hashTo === null) {
    hashTo = unhashedType.length;
  }
  //for (let i = hashes[0][0]; i < hashTo; i++) {
  //  const hash = hashes[i][1];
  //  const action = unhashedType[i];
  //  const newAction = Object.assign(
  //    Object.create(Object.getPrototypeOf(action)),
  //    action
  //  );
  //  newAction.hash = hash
  //  hashed[i] = newAction;
  //}

  for (let i = 0; i < hashTo; i++) {
    const action = hashed[i];
    if (action.hash === null) {
      for (let [idx, hash] of hashes) {
        if (idx === i) {
          const newAction = Object.assign(
            Object.create(Object.getPrototypeOf(action)),
            action
          );
          newAction.hash = hash
          hashed[i] = newAction;
        }
      }
    }
  }

  return hashed;
}
