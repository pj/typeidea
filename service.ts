type ServiceState = 'active' | 'deprecated' | 'remove';

export class ServiceAction {
  changeLog: string;
  hash: string | null;
  state: ServiceState;

  constructor(changeLog: string, hash: string | null, state: ServiceState) {
    this.changeLog = changeLog;
    this.hash = hash;
    this.state = state;
  }

  fieldsToHash(): string {
    throw new Error('NotImplemented');
  };
}

export class NewServiceAction extends ServiceAction {
  description: string;
  name: string;

  constructor(
    changeLog: string,
    hash: string | null,
    state: ServiceState,
    description: string,
    name: string
  ) {
    super(changeLog, hash, state);
    this.description = description;
    this.name = name;
  }

  fieldsToHash(): string {
    return `${this.changeLog}`;
  };
}

export class UpdateDescriptionAction extends ServiceAction {
  description: string;

  constructor(changeLog: string, hash: string | null, state: ServiceState, description: string) {
    super(changeLog, hash, state);
    this.description = description;
  }

  fieldsToHash(): string {
    return `${this.changeLog}`;
  };
}

export class AddInputVersionAction extends ServiceAction {
  constructor(changeLog: string, hash: string | null, state: ServiceState) {
    super(changeLog, hash, state);
  }

  fieldsToHash(): string {
    return `${this.changeLog}`;
  };
}

export class RemoveInputVersionAction extends ServiceAction {
  constructor(changeLog: string, hash: string | null, state: ServiceState) {
    super(changeLog, hash, state);
  }

  fieldsToHash(): string {
    return `${this.changeLog}`;
  };
}

export class DeprecateInputVersionAction extends ServiceAction {
  constructor(changeLog: string, hash: string | null, state: ServiceState) {
    super(changeLog, hash, state);
  }

  fieldsToHash(): string {
    return `${this.changeLog}`;
  };
}

export class AddOutputVersionAction extends ServiceAction {
  constructor(changeLog: string, hash: string | null, state: ServiceState) {
    super(changeLog, hash, state);
  }

  fieldsToHash(): string {
    return `${this.changeLog}`;
  };
}

export class RemoveOutputVersionAction extends ServiceAction {
  constructor(changeLog: string, hash: string | null, state: ServiceState) {
    super(changeLog, hash, state);
  }

  fieldsToHash(): string {
    return `${this.changeLog}`;
  };
}

export class DeprecateOutputVersionAction extends ServiceAction {
  constructor(changeLog: string, hash: string | null, state: ServiceState) {
    super(changeLog, hash, state);
  }

  fieldsToHash(): string {
    return `${this.changeLog}`;
  };
}
