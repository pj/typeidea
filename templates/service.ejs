/**
* **GENERATED CODE DO NOT EDIT!**
*
* <%= service.name %>
*
* <%= service.description %>
*
* # Change Log:
<% service.changeLog.forEach(function (log, index) { -%>
* <%= index + 1 %>. <%= log %>
<%_ }) -%>
*/

import * as <%= service.inputType %> from './<%= service.inputType %>';
import * as <%= service.outputType %> from './<%= service.outputType %>';

import * as express from 'express';
import * as bodyParser from 'body-parser';

const app = express();
app.use(bodyParser.json());

function <%= service.name %>(
func: (
<%= service.inputVersions.map(function (inputVersion) { return inputVersion.version } ).join(' | ') %>
) => <%= service.outputVersions.map(function (outputVersion) { return outputVersion.version } ).join(' | ') %>
): void {
  app.get('/<%= service.name %>', (req, res) => {
    const inputMessage = <%= service.inputType %>.deserialize(req.body);
    const response = func(inputMessage);
    const outputMessage = <%= service.outputType %>.serialize(response);
    res.json(outputMessage);
  });
}

export function startServices() {
  app.listen(3000, () => console.log(`Example app listening on port ${port}!`));
}
