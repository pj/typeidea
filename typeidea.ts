import * as crypto from 'crypto';
import {compile} from 'ejs';
import * as fs from 'fs';
import * as prettier from 'prettier';

const typescriptTypeFile = fs.readFileSync(
  './templates/typescript.ejs',
  {
    encoding: "utf8",
  }
);
const typescriptTypeTemplate = compile(typescriptTypeFile);

type FieldTypes = 'string' | 'boolean' | 'number';

type FieldDefaults = string | boolean | number;

type Hash = string | null;

class Action {
  changeLog: string;
  hash: string | null;

  constructor(changeLog: string, hash: string | null) {
    this.changeLog = changeLog;
    this.hash = hash;
  }

  fieldsToHash(): string {
    return `${this.changeLog}`;
  };
}

export class NewAction extends Action {
  name: string;

  constructor(changeLog: string, hash: string | null, name: string) {
    super(changeLog, hash);
    this.name = name;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.name}`;
  };
}

export class RenameAction extends Action {
  _from: string;
  to: string;

  constructor(changeLog: string, hash: string | null, _from: string, to: string) {
    super(changeLog, hash);
    this._from = _from;
    this.to = to;
  }
}

export class RequiredAction extends Action {
  name: string;

  constructor(changeLog: string, hash: string | null, name: string) {
    super(changeLog, hash);
    this.name = name;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.name}`;
  };
}

export class OptionalAction extends Action {
  name: string;

  constructor(changeLog: string, hash: string | null, name: string) {
    super(changeLog, hash);
    this.name = name;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.name}`;
  };
}

export class DeleteAction extends Action {
  name: string;

  constructor(changeLog: string, hash: string | null, name: string) {
    super(changeLog, hash);
    this.name = name;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.name}`;
  };
}

export class SetDefaultAction extends Action {
  name: string;
  _default: FieldDefaults;

  constructor(changeLog: string, hash: string | null, name: string, _default: FieldDefaults) {
    super(changeLog, hash);
    this.name = name;
    this._default = _default;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.name}_${this._default}`;
  };
}

export class RemoveDefaultAction extends Action {
  name: string;

  constructor(changeLog: string, hash: string | null, name: string) {
    super(changeLog, hash);
    this.name = name;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.name}`;
  };
}

export class AddAction extends Action {
  name: string;
  type: FieldTypes;
  description: string;
  optional: boolean;
  _default: FieldDefaults | null;

  constructor(
    changeLog: string, hash: string | null, name: string, type: FieldTypes,
    description: string, optional: boolean, _default: FieldDefaults | null
  ) {
    super(changeLog, hash);
    this.name = name;
    this.type = type;
    this.description = description;
    this.optional = optional;
    this._default = _default;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.name}_${this.type}_${this.description}_${this.optional}_${this._default}`;
  };
}

export class UpdateDescriptionAction extends Action {
  name: string;
  description: string;

  constructor(changeLog: string, hash: string | null, name: string, description: string) {
    super(changeLog, hash);
    this.name = name;
    this.description = description;
  }
}

export class ReferenceAction extends Action {
  name: string;
  description: string;
  optional: boolean;
  referenceType: string;
  referenceHash: string;

  constructor(
    changeLog: string, hash: string | null, name: string, description: string,
    optional: boolean, referenceType: string, referenceHash: string
  ) {
    super(changeLog, hash);
    this.name = name;
    this.description = description;
    this.optional = optional;
    this.referenceType = referenceType;
    this.referenceHash = referenceHash;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.name}_${this.description}_${this.optional}_${this.referenceType}_${this.referenceHash}`;
  };
}
// services

function service(name: string) {

}

function hashAction(action: Action, previous: string | null) {
  const hash = crypto.createHash('sha256');

  if (previous !== null) {
    hash.update(previous);
  }

  const fieldsToHash = action.fieldsToHash();
  hash.update(fieldsToHash);

  return hash.digest('hex');
}

/*
Take a list of types and services and verify the types and generate any empty
types.
 */
export function hashTypes(
  types: Array<Array<Action>>,
  services: any
): Array<Array<[number, string]>> {
  const hashedTypes: Array<Array<[number, string]>> = [];
  for (const type of types) {
    const newHashes: Array<[number, string]> = [];
    const first = type[0];
    if (!(first instanceof NewAction)) {
      throw new Error(`Type must start with a NewAction element ${first}`);
    }
    let previousHash = hashAction(first, null);
    if (first.hash === null) {
      newHashes.push([0, previousHash]);
    }
    if (first.hash !== null && previousHash !== first.hash) {
      throw new Error(`Invalid hash at item ${0} ${first}, did you change something?`)
    }
    let createHashes = false;

    for (let n = 1; n < type.length; n++) {
      let action = type[n];
      if (createHashes) {
        if (action.hash !== null) {
          throw new Error(`Hash entered when there were previous items without hashes index: ${n} action: ${action}`);
        }
        previousHash = hashAction(action, previousHash);
        newHashes.push([n, previousHash]);
      } else {
        if (action.hash === null) {
          previousHash = hashAction(action, previousHash);
          newHashes.push([n, previousHash]);
          createHashes = true;
        } else {
          const expectedHash = hashAction(action, previousHash);
          if (expectedHash !== action.hash) {
            throw new Error(`Invalid hash at item ${n} ${action}, did you change something?`)
          }
          previousHash = action.hash;
        }
      }
    }

    hashedTypes.push(newHashes);
  }

  return hashedTypes;
}

export class BaseField {
  name: string;
  changeLog: string;
  description: string;
  optional: boolean;

  constructor(
    name: string,
    changeLog: string,
    description: string,
    optional: boolean,
  ) {
    this.name = name;
    this.changeLog = changeLog;
    this.description = description;
    this.optional = optional;
  }

  fieldType(): string {
    throw new Error("Not implemented");
  }

  copy(): BaseField {
    throw new Error("Not implemented");
  }
}

export class Field extends BaseField {
  type: FieldTypes;
  _default: FieldDefaults | null;

  constructor(
    name: string,
    changeLog: string,
    description: string,
    optional: boolean,
    type: FieldTypes,
    _default: FieldDefaults | null
  ) {
    super(name, changeLog, description, optional);
    this.type = type;
    this._default = _default;
  }

  copy(): Field {
    return new Field(
      this.name,
      this.changeLog,
      this.description,
      this.optional,
      this.type,
      this._default
    );
  }

  fieldType(): string {
    return this.type + (this.optional ? " | null" : "");
  }

  formattedDefault(): string {
    if (!this._default) {
      return "";
    }
    if (this.type === 'string') {
      return `"${this._default}"`;
    }
    return "" + this._default;
  }
}

export class ReferenceField extends BaseField {
  referenceType: string;
  referenceHash: string;

  constructor(
    name: string,
    changeLog: string,
    description: string,
    optional: boolean,
    referenceType: string,
    referenceHash: string
  ) {
    super(name, changeLog, description, optional);
    this.referenceType = referenceType;
    this.referenceHash = referenceHash;
  }

  copy() {
    return new ReferenceField(
      this.name,
      this.changeLog,
      this.description,
      this.optional,
      this.referenceType,
      this.referenceHash
    );
  }

  fieldType() {
    return `${this.referenceType}.h_${this.referenceHash}`;
  }

  formattedDefault(): string {
    return "";
  }
}

type FieldObject = {
  [key: string]: BaseField;
};

export class Version {
  hash: string | null;
  fields: FieldObject;

  constructor(hash: string | null, fields: FieldObject) {
    this.hash = hash;
    this.fields = fields;
  }
}

export class Type {
  name: string;
  versions: Version[];

  constructor(name: string) {
    this.name = name;
    this.versions = [];
  }
}

/*
Generate types and services (and verify).
 */
export function generateTypes(types: Array<Array<Action>>, services: any): Type[] {
  const generatedTypes = [];
  for (const type of types) {
    const versions = [];
    for (let n = 0; n < type.length; n++) {
      const action = type[n];
      if (action.hash === null) {
        throw new Error(`No hash at ${n} ${action}, did you forget to add it?`)
      }
    }

    const first = (<NewAction>type[0]);
    if (!(first instanceof NewAction)) {
      throw new Error(`Type must start with a NewAction element ${first}`);
    }
    let previousHash = hashAction(first, null);
    if (previousHash !== first.hash) {
      throw new Error(`Invalid hash at item ${0} ${first}, did you change something?`)
    }
    let createHashes = false;

    for (let n = 1; n < type.length; n++) {
      let action = type[n];
      const expectedHash = hashAction(action, previousHash);
      if (expectedHash !== action.hash) {
        throw new Error(`Invalid hash at item ${n} ${action}, did you change something?`)
      }
      previousHash = action.hash;
    }

    const name = first.name;
    let previousVersion = null;
    const newType = new Type(name);

    for (let n = 1; n < type.length; n++) {
      let action = type[n];
      const newVersion = new Version(action.hash, {});
      if (previousVersion) {
        newVersion.fields = {...previousVersion.fields}
      }

      if (action instanceof RenameAction) {
        const currentField = newVersion.fields[action._from];
        const newField = currentField.copy();
        newField.name = action.to;
        newField.changeLog = action.changeLog;
        delete newVersion.fields[action._from];
        newVersion.fields[action.to] = newField;
      } else if (action instanceof RequiredAction) {
        const currentField = newVersion.fields[action.name];
        const newField = currentField.copy();
        newField.optional = false;
        newField.changeLog = action.changeLog;
        newVersion.fields[newField.name] = newField;
      } else if (action instanceof OptionalAction) {
        const currentField = newVersion.fields[action.name];
        const newField = currentField.copy();
        newField.optional = true;
        newField.changeLog = action.changeLog;
        newVersion.fields[newField.name] = newField;
      } else if (action instanceof DeleteAction) {
        const currentField = newVersion.fields[action.name];
        const newField = currentField.copy();
        delete newVersion.fields[currentField.name];
      } else if (action instanceof SetDefaultAction) {
        const currentField = newVersion.fields[action.name];
        const newField = currentField.copy();
        if (newField instanceof Field) {
          newField._default = action._default;
        }
        newField.changeLog = action.changeLog;
        newVersion.fields[newField.name] = newField;
      } else if (action instanceof RemoveDefaultAction) {
        const currentField = newVersion.fields[action.name];
        const newField = currentField.copy();
        if (newField instanceof Field) {
          newField._default = null;
        }
        newField.changeLog = action.changeLog;
        newVersion.fields[newField.name] = newField;
      } else if (action instanceof AddAction) {
        const currentField = newVersion.fields[action.name];
        const newField = new Field(
          action.name,
          action.changeLog,
          action.description,
          action.optional,
          action.type,
          action._default
        );
        newVersion.fields[newField.name] = newField;
      } else if (action instanceof UpdateDescriptionAction) {
        const currentField = newVersion.fields[action.name];
        const newField = currentField.copy();
        newField.description = action.description;
        newField.changeLog = action.changeLog;
        newVersion.fields[newField.name] = newField;
      } else if (action instanceof ReferenceAction) {
        const currentField = newVersion.fields[action.name];
        const newField = new ReferenceField(
          action.name,
          action.changeLog,
          action.description,
          action.optional,
          action.referenceType,
          action.referenceHash
        );
        newVersion.fields[newField.name] = newField;
      }

      newType.versions.push(newVersion);
      previousVersion = newVersion;
    }
    generatedTypes.push(newType);
  }

  return generatedTypes;
}

export function generateTypescript(types: Type[]) {
  return types.map((_type) => {
    // load imports.
    const imports = new Set();
    for (const version of _type.versions) {
      for (const field of Object.values(version.fields)) {
        if (field instanceof ReferenceField) {
          imports.add(field.referenceType);
        }
      }
    }
    return [
      _type,
      prettier.format(typescriptTypeTemplate({versions: _type.versions, imports: imports}), { parser: 'typescript' })
    ];
  }
  );
}

export function addHashes(
  unhashedType: Action[],
  hashes: Array<[number, string]>
) {
  const hashed = unhashedType.slice();
  for (let i = hashes[0][0]; i < unhashedType.length; i++) {
    const hash = hashes[i][1];
    const action = unhashedType[i];
    const newAction = Object.assign(
      Object.create(Object.getPrototypeOf(action)),
      action
    );
    newAction.hash = hash
    hashed[i] = newAction;
  }

  return hashed;
}

export function loadActions(path: string): Array<Array<Action>> {
  const types = require(path);
  const outputTypes = [];

  for (const _type of types) {
    const log = [];
    for (const action of _type) {
      switch(action._action_type) {
        case 'RenameAction':
          log.push(new RenameAction(action.changeLog, action.hash, action._from, action.to));
          break;
        case 'RequiredAction':
          log.push(new RequiredAction(action.changeLog, action.hash, action.name));
          break;
        case 'OptionalAction':
          log.push(new OptionalAction(action.changeLog, action.hash, action.name));
          break;
        case 'DeleteAction':
          log.push(new DeleteAction(action.changeLog, action.hash, action.name));
          break;
        case 'SetDefaultAction':
          log.push(new SetDefaultAction(action.changeLog, action.hash, action.name, action._default));
          break;
        case 'RemoveDefaultAction':
          log.push(new RemoveDefaultAction(action.changeLog, action.hash, action.name));
          break;
        case 'AddAction':
          log.push(new AddAction(action.changeLog, action.hash, action.name, action.type, action.description, action.optional, action._default));
          break;
        case 'UpdateDescriptionAction':
          log.push(new UpdateDescriptionAction(action.changeLog, action.hash, action.name, action.description));
          break;
        case 'ReferenceAction':
          log.push(new ReferenceAction(action.changeLog, action.hash, action.name, action.description, action.optional, action.referenceType, action.referenceHash));
          break;
        case 'NewAction':
          log.push(new NewAction(action.changeLog, action.hash, action.name));
          break;
      }
    }
    outputTypes.push(log);
  }

  return outputTypes;
}
