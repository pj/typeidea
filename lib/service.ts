import * as typeidea from './typeidea';
import {
  Action,
  AddInputVersionAction,
  RemoveInputVersionAction,
  DeprecateInputVersionAction,
  AddOutputVersionAction,
  NewServiceAction,
  DeprecateOutputVersionAction,
  RemoveOutputVersionAction,
  UpdateDescriptionAction
} from './action';
import {compile} from 'ejs';
import * as fs from 'fs';
import * as prettier from 'prettier';

const typescriptServiceFile = fs.readFileSync(
  './templates/service.ejs',
  {
    encoding: "utf8",
  }
);
const typescriptServiceTemplate = compile(
  typescriptServiceFile,
  {
    filename: './templates/service.ejs'
  }
);

//export class ServiceAction {
//  changeLog: string;
//  hash: string | null;
//
//  constructor(changeLog: string, hash: string | null) {
//    this.changeLog = changeLog;
//    this.hash = hash;
//  }
//
//  fieldsToHash(): string {
//    throw new Error('NotImplemented');
//  };
//}
//
//export class NewServiceAction extends ServiceAction {
//  name: string;
//  description: string;
//  inputType: string;
//  outputType: string;
//  inputVersion: string;
//  outputVersion: string;
//
//  constructor(
//    changeLog: string,
//    hash: string | null,
//    name: string,
//    description: string,
//    inputType: string,
//    outputType: string,
//    inputVersion: string,
//    outputVersion: string,
//  ) {
//    super(changeLog, hash);
//    this.name = name;
//    this.description = description;
//    this.inputType = inputType;
//    this.outputType = outputType;
//    this.inputVersion = inputVersion;
//    this.outputVersion = outputVersion;
//  }
//
//  fieldsToHash(): string {
//    return `${this.changeLog}_${this.name}_${this.description}_${this.inputType}_${this.outputType}_${this.inputVersion}_${this.outputVersion}`;
//  };
//}
//
//export class UpdateDescriptionAction extends ServiceAction {
//  description: string;
//
//  constructor(
//    changeLog: string,
//    hash: string | null,
//    description: string,
//  ) {
//    super(changeLog, hash);
//    this.description = description;
//  }
//
//  fieldsToHash(): string {
//    return `${this.changeLog}_${this.description}`;
//  };
//}
//
//export class AddInputVersionAction extends ServiceAction {
//  version: string;
//
//  constructor(
//    changeLog: string,
//    hash: string | null,
//    version: string,
//  ) {
//    super(changeLog, hash);
//    this.version = version;
//  }
//
//  fieldsToHash(): string {
//    return `${this.changeLog}_${this.version}`;
//  };
//}
//
//export class RemoveInputVersionAction extends ServiceAction {
//  version: string;
//
//  constructor(
//    changeLog: string,
//    hash: string | null,
//    version: string,
//  ) {
//    super(changeLog, hash);
//    this.version = version;
//  }
//
//  fieldsToHash(): string {
//    return `${this.changeLog}_${this.version}`;
//  };
//}
//
//export class DeprecateInputVersionAction extends ServiceAction {
//  version: string;
//
//  constructor(
//    changeLog: string,
//    hash: string | null,
//    version: string,
//  ) {
//    super(changeLog, hash);
//    this.version = version;
//  }
//
//  fieldsToHash(): string {
//    return `${this.changeLog}_${this.version}`;
//  };
//}
//
//export class AddOutputVersionAction extends ServiceAction {
//  version: string;
//
//  constructor(
//    changeLog: string,
//    hash: string | null,
//    version: string,
//  ) {
//    super(changeLog, hash);
//    this.version = version;
//  }
//
//  fieldsToHash(): string {
//    return `${this.changeLog}_${this.version}`;
//  };
//}
//
//export class RemoveOutputVersionAction extends ServiceAction {
//  version: string;
//
//  constructor(
//    changeLog: string,
//    hash: string | null,
//    version: string,
//  ) {
//    super(changeLog, hash);
//    this.version = version;
//  }
//
//  fieldsToHash(): string {
//    return `${this.changeLog}_${this.version}`;
//  };
//}
//
//export class DeprecateOutputVersionAction extends ServiceAction {
//  _type: string;
//  version: string;
//
//  constructor(
//    changeLog: string,
//    hash: string | null,
//    version: string,
//  ) {
//    super(changeLog, hash);
//    this.version = version;
//  }
//
//  fieldsToHash(): string {
//    return `${this.changeLog}_${this.version}`;
//  };
//}
//
type ServiceState = 'active' | 'deprecated' | 'removed';

export class ServiceVersion {
  version: string;
  state: ServiceState;

  constructor(
    version: string,
    state: ServiceState,
  ) {
    this.version = version;
    this.state = state;
  }
}

export class Service {
  name: string;
  changeLog: string[];
  description: string;
  inputType: string;
  outputType: string;
  inputVersions: ServiceVersion[];
  outputVersions: ServiceVersion[];

  constructor(
    name: string,
    description: string,
    inputType: string,
    outputType: string,
  ) {
    this.name = name;
    this.description = description;
    this.changeLog = [];
    this.inputType = inputType;
    this.outputType = outputType;
    this.inputVersions = [];
    this.outputVersions = [];
  }
}

export function generateServices(services: Array<Array<Action>>): Service[] {
  const generatedServices = [];
  for (const service of services) {
    let previousHash = null;
    const changeLog = [];
    let notHashed = false;
    for (let n = 0; n < service.length; n++) {
      const action = service[n];
      if (notHashed) {
        if (action.hash !== null) {
          throw new Error(`Hashed action after unhashed action at ${n} ${action}`);
        }
      } else if (action.hash === null) {
        notHashed = true;
      } else {
        const expectedHash = typeidea.hashAction(action, previousHash);
        if (expectedHash !== action.hash) {
          throw new Error(`Invalid hash at item ${n} ${action}, did you change something?`)
        }
        previousHash = expectedHash;
      }
    }

    const newAction = (service[0] as NewServiceAction);
    const newService = new Service(
      newAction.name,
      newAction.description,
      newAction.inputType,
      newAction.outputType,
    );
    newService.inputVersions.push(new ServiceVersion(newAction.inputVersion, 'active'));
    newService.outputVersions.push(new ServiceVersion(newAction.outputVersion, 'active'));

    for (let n = 1; n < service.length; n++) {
      let action = service[n];
      if (action instanceof NewServiceAction) {
          throw new Error(`New Service action not at start!`);
      } else if (action instanceof UpdateDescriptionAction) {
          newService.description = action.description;
      } else if (action instanceof AddInputVersionAction) {
          newService.inputVersions.push(
            new ServiceVersion(action.version, 'active')
          );
      } else if (action instanceof RemoveInputVersionAction) {
        newService.inputVersions = newService.inputVersions.map(serviceVersion => {
          if (serviceVersion.version === (action as any).version) {
            return new ServiceVersion((action as any).version, 'removed');
          }
          return serviceVersion;
        });
      } else if (action instanceof DeprecateInputVersionAction) {
        newService.inputVersions = newService.inputVersions.map(serviceVersion => {
          if (serviceVersion.version === (action as any).version) {
            return new ServiceVersion((action as any).version, 'deprecated');
          }
          return serviceVersion;
        });
      } else if (action instanceof AddOutputVersionAction) {
        newService.outputVersions.push(
          new ServiceVersion(action.version, 'active')
        );
      } else if (action instanceof RemoveOutputVersionAction) {
        newService.outputVersions = newService.outputVersions.map(serviceVersion => {
          if (serviceVersion.version === (action as any).version) {
            return new ServiceVersion((action as any).version, 'removed');
          }
          return serviceVersion;
        });
      } else if (action instanceof DeprecateOutputVersionAction) {
        newService.outputVersions = newService.outputVersions.map(serviceVersion => {
          if (serviceVersion.version === (action as any).version) {
            return new ServiceVersion((action as any).version, 'deprecated');
          }
          return serviceVersion;
        });
      }

      newService.changeLog.push(action.changeLog);
    }
    generatedServices.push(newService);
  }

  return generatedServices;
}

export function generateTypescript(services: Service[]) {
  return services.map((service) => {
    return [
      service,
      prettier.format(
        typescriptServiceTemplate(
          {
            service: service
          }
        ),
        {parser: 'typescript'},
      )
    ];
  }
  );
}
function createActions(actions: any[]): Action[] {
  const log = [];

  for (const action of actions) {
    switch(action._action_type) {
      case 'NewServiceAction':
        log.push(new NewServiceAction(
          action.changeLog,
          action.hash,
          action.name,
          action.description,
          action.inputType,
          action.outputType,
          action.inputVersion,
          action.outputVersion,
        )
        );
        break;
      case 'UpdateDescriptionAction':
        log.push(new UpdateDescriptionAction(
          action.changeLog,
          action.hash,
          action.description)
        );
        break;
      case 'AddInputVersionAction':
        log.push(new AddInputVersionAction(
          action.changeLog,
          action.hash,
          action.version)
        );
        break;
      case 'RemoveInputVersionAction':
        log.push(new RemoveInputVersionAction(
          action.changeLog,
          action.hash,
          action.version)
        );
        break;
      case 'DeprecateInputVersionAction':
        log.push(new DeprecateInputVersionAction(
          action.changeLog,
          action.hash,
          action.version
        ));
        break;
      case 'AddOutputVersionAction':
        log.push(new AddOutputVersionAction(
          action.changeLog,
          action.hash,
          action.version
        ));
        break;
      case 'RemoveOutputVersionAction':
        log.push(new RemoveOutputVersionAction(
          action.changeLog,
          action.hash,
          action.version
        ));
        break;
      case 'DeprecateOutputVersionAction':
        log.push(new DeprecateOutputVersionAction(
          action.changeLog,
          action.hash,
          action.version
        ));
        break;
    }
  }

  return log;
}

export function loadActions(path: string): Array<Array<Action>> {
  const types = require(path);
  const outputTypes = [];

  for (const _type of types) {
    const log = createActions(_type);
    outputTypes.push(log);
  }

  return outputTypes;
}
