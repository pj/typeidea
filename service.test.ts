import * as typeidea from './typeidea';
import * as service from './service';

const json_snapshot_tests = ([
  ['Create a service', './tests/services/create_service.json', null],
] as Array<[string, string, number | null]>);

for (const [name, path, hashTo] of json_snapshot_tests) {
  it(name, () => {
    const addField = service.loadActions(path);

    const hashes = typeidea.hashActions(addField);
    const hashedAddField = typeidea.addHashes(addField[0], hashes[0], hashTo);

    const services = service.generateServices([hashedAddField]);
    const typescript = service.generateTypescript(services);

    expect(typescript[0][1]).toMatchSnapshot();
  });
}

