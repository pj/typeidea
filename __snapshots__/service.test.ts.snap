// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create a service 1`] = `
"/**
 * **GENERATED CODE DO NOT EDIT!**
 *
 * TestService
 *
 * This is a Test service
 *
 * # Change Log:
 */

import * as TestInputMessage from \\"./TestInputMessage\\";
import * as TestOutputMessage from \\"./TestOutputMessage\\";

import * as express from \\"express\\";
import * as bodyParser from \\"body-parser\\";

const app = express();
app.use(bodyParser.json());

function TestService(func: (V0) => V0): void {
  app.get(\\"/TestService\\", (req, res) => {
    const inputMessage = TestInputMessage.deserialize(req.body);
    const response = func(inputMessage);
    const outputMessage = TestOutputMessage.serialize(response);
    res.json(outputMessage);
  });
}

export function startServices() {
  app.listen(3000, () => console.log(\`Example app listening on port \${port}!\`));
}
"
`;
